simulated_beacon/simulated_beacon.go

```diff
@@ -0,0 +1,236 @@
+// Copyright 2023 The go-ethereum Authors
+// This file is part of the go-ethereum library.
+//
+// The go-ethereum library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Lesser General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// The go-ethereum library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the go-ethereum library. If not, see <http://www.gnu.org/licenses/>.
+
+package simulated_beacon
+
+import (
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/ethereum/go-ethereum/beacon/engine"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/ethereum/go-ethereum/eth"
+	"github.com/ethereum/go-ethereum/eth/catalyst"
+	"github.com/ethereum/go-ethereum/log"
+	"github.com/ethereum/go-ethereum/node"
+	"github.com/ethereum/go-ethereum/rpc"
+)
+
+// withdrawals implements a FIFO queue which holds withdrawals that are pending inclusion
+type withdrawals struct {
+	pending []*types.Withdrawal
+	next    uint64
+	mu      sync.Mutex
+}
+
+// pop removes up to 10 withdrawals from the queue
+func (w *withdrawals) pop() []*types.Withdrawal {
+	w.mu.Lock()
+	defer w.mu.Unlock()
+
+	var popCount int
+	if len(w.pending) >= 10 {
+		popCount = 10
+	} else {
+		popCount = len(w.pending)
+	}
+
+	popped := make([]*types.Withdrawal, popCount)
+	copy(popped[:], w.pending[0:popCount])
+	w.pending = append([]*types.Withdrawal{}, w.pending[popCount:]...)
+	return popped
+}
+
+// add adds a withdrawal to the queue
+func (w *withdrawals) add(withdrawal *types.Withdrawal) error {
+	w.mu.Lock()
+	defer w.mu.Unlock()
+
+	if withdrawal.Index < w.next {
```

Comment:
I kinda think you should just auto-assign the index. There is technically no index verification on the EL and this already ensures the index in monotonically increasing.

Reply:
seems fine

Reply:
How do I set the index to automatic? Why is my system set to zero? I have a lot of assets. Why should it be like this? Please guide me?